lib/gnat.ex

```diff
@@ -149,17 +168,29 @@ defmodule Gnat do
         {:reply, :ok, state}
     end
   end
+  def handle_call({:ping, pinger}, _from, state) do
+    :ok = :gen_tcp.send(state.tcp, "PING\r\n")
+    {:reply, :ok, Map.put(state, :pinger, pinger)}
+  end
 
-  defp perform_handshake(tcp) do
+  defp perform_handshake(tcp, connection_settings) do
     receive do
       {:tcp, ^tcp, operation} ->
-        "INFO" = operation |> String.split |> List.first |> String.upcase
-        :gen_tcp.send(tcp, "CONNECT {\"verbose\": false}\r\n")
+        {_, [{:info, options}]} = Parser.parse(Parser.new, operation)
+        connect(tcp, options, connection_settings)
       after 1000 ->
         {:error, "timed out waiting for info"}
     end
   end
 
+  defp connect(tcp, %{auth_required: true}=_options, %{username: username, password: password}=_connection_settings) do
```

Comment:
I don't think we need to assign unused variables `_connection_settings` and `_options` here. The pattern matching should be just fine and if someone wanted context about _what_ is being matched, they can look and the other `connect` function